lib/src/multipart_request.dart

```diff
@@ -4,162 +4,143 @@
-  /// request body. According to http://tools.ietf.org/html/rfc1341.html, this
-  /// can't be longer than 70.
-  static const int _BOUNDARY_LENGTH = 70;
-
-  static final Random _random = new Random();
-
-  /// The form fields to send for this request.
-  final Map<String, String> fields;
-
-  /// The private version of [files].
-  final List<MultipartFile> _files;
-
-  /// Creates a new [MultipartRequest].
-  MultipartRequest(String method, Uri url)
-    : fields = {},
-      _files = <MultipartFile>[],
-      super(method, url);
-
-  /// The list of files to upload for this request.
-  List<MultipartFile> get files => _files;
-
-  /// The total length of the request body, in bytes. This is calculated from
-  /// [fields] and [files] and cannot be set manually.
-  int get contentLength {
-    var length = 0;
-
-    fields.forEach((name, value) {
-      length += "--".length + _BOUNDARY_LENGTH + "\r\n".length +
-          UTF8.encode(_headerForField(name, value)).length +
-          UTF8.encode(value).length + "\r\n".length;
-    });
-
-    for (var file in _files) {
-      length += "--".length + _BOUNDARY_LENGTH + "\r\n".length +
-          UTF8.encode(_headerForFile(file)).length +
-          file.length + "\r\n".length;
-    }
-
-    return length + "--".length + _BOUNDARY_LENGTH + "--\r\n".length;
-  }
-
-  void set contentLength(int value) {
-    throw new UnsupportedError("Cannot set the contentLength property of "
-        "multipart requests.");
-  }
-
-  /// Freezes all mutable fields and returns a single-subscription [ByteStream]
-  /// that will emit the request body.
-  ByteStream finalize() {
-    // TODO(nweiz): freeze fields and files
-    var boundary = _boundaryString();
-    headers['content-type'] = 'multipart/form-data; boundary=$boundary';
-    super.finalize();
-
+class MultipartBody implements Body {
+  /// The contents of the message body.
+  ///
+  /// This will be `null` after [read] is called.
+  Stream<List<int>> _stream;
+
+  /// The length of the stream returned by [read].
+  ///
+  /// This is calculated from the fields and files passed into the body.
+  final int contentLength;
+
+  /// Multipart forms do not have an encoding.
+  Encoding get encoding => null;
+
+  /// Creates a [MultipartBody] from the given [fields] and [files].
+  ///
+  /// The [boundary] is used to
+  factory MultipartBody(
+    Map<String, String> fields,
+    List<MultipartFile> files,
+    String boundary,
+  ) {
     var controller = new StreamController<List<int>>(sync: true);
+    var contentLength = 0;
 
     void writeAscii(String string) {
-      controller.add(UTF8.encode(string));
+      controller.add(string.codeUnits);
+      contentLength += string.length;
+    }
+    void writeUtf8(String string) {
+      var encoded = UTF8.encode(string);
+      controller.add(encoded);
+      contentLength += encoded.length;
+    }
+    void writeLine() {
+      controller.add([13, 10]); // \r\n
+      contentLength += 2;
```

Comment:
Since you're only using these for fields now, you might as well just add everything to a single `Uint8List` rather than adding separate chunks to a controller. The `typed_data` package can help with that.